### PR TITLE
Add condition for clippy to warn on naked unwraps

### PR DESCRIPTION
**Pull request by njgheorghita**
### What was wrong?
Naked unwraps are bad, and have a tendency slip through the cracks. Case in point, I missed one in #555. Even though @ogenev spotted it, I think it's a good idea to add a check for unwraps to clippy so we don't have to rely on humans.

Furthermore, while working with the testnet nodes, I've experienced multiple instances where nodes are crashing / behaving incorrectly due to these unwraps. 
- eg. I spent a solid day once trying to figure out why a node was running but wouldn't respond to `portal_historyPaginate*` endpoints. Turns out the history thread was crashing due to a naked unwrap.

Our style guide also [dictates](https://github.com/ethereum/trin/blob/master/docs/contributing.md#error-handling) that naked unwraps are not allowed and if there is an exception (eg parsing a static value), the justification must be labeled.

In this pr, I added a clippy check for `unwrap_used`. I went through and fixed all of the simple cases. I plan to go through (before merging this pr) and revisit the more complex cases once I get :+1: on adopting this lint from the others

### How was it fixed?
- Added `#[warn(clippy::unwrap_used)]` to each workspace, except...
  - `trin-cli` - this library is no longer "maintained"
  - `utp-testing` - since it's a "testing" library
- Introduced a `clippy.toml` file to allow for `allow-unwrap-in-tests` configuration 
- Currently, the `allow-unwrap-in-tests` lint is [only available in rust nightly, not stable](https://github.com/rust-lang/rust-clippy/issues/9062). Which requires an explicit `#[allow(clippy::unwrap_used)]` for each test block that contains unwraps - since our ci environment uses stable. Imo, this is a bit wordy, but a fair trade-off since...
  - we'll be able to remove these explicit `allows` once the feature gets in stable
  - according to the [clippy docs](https://doc.rust-lang.org/clippy/usage.html?highlight=meant#lint-configuration)
    - `Clippy is meant to be used with a generous sprinkling of #[allow(..)]s through your code.`
  - in the meantime we get the benefit of automatically checking for naked unwraps

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

**Comment by carver** [comment]
> In this pr, I added a clippy check for `unwrap_used`. I went through and fixed all of the simple cases. I plan to go through (before merging this pr) and revisit the more complex cases once I get +1 on adopting this lint from the others

So right now you're just looking on +1 to adopt the lint, then you'll come back and ask for another review later?

If so, then 👍🏻 for adding this lint.

**Comment by mrferris** [comment]
👍 from me as well

**Comment by ogenev** [comment]
:+1:

**Comment by njgheorghita** [comment]
@carver @mrferris @ogenev ready for an :eyes: on this one

File: trin-core/src/portalnet/overlay_service.rs [hunk 1]
```diff
@@ -1302,10 +1302,19 @@ where
...
+                Err(err) => {
+                    error!(error = %err, "Error receiving UTP stream from UtpListener");
+                    return;
+                }
+            };
             // Handle STATE packet for SYN
             let mut buf = [0; BUF_SIZE];
-            conn.recv(&mut buf).await.unwrap();
+            if let Err(err) = conn.recv(&mut buf).await {
+                error!(error = %err, "Error receiving UTP stream from UtpListener");
```
Comment (carver) [inline trin-core/src/portalnet/overlay_service.rs:1315]: It's almost always helpful to have the string be unique. To make this different from 1308:
```suggestion
                error!(error = %err, "Error receiving STATE packet from UtpListener");
```

File: trin-core/src/portalnet/storage.rs [hunk 2]
```diff
@@ -427,7 +431,8 @@ impl PortalStorage {
 
         // Delete furthest data until our data usage is less than capacity.
         while self.capacity_reached()? {
-            // Unwrap because this was set in the block before this loop.
+            // Naked unwrap ok because this was set in the block before this loop.
+            #[allow(clippy::unwrap_used)]
             let id_to_remove = self.farthest_content_id.clone().unwrap();
```
Comment (carver) [inline trin-core/src/portalnet/storage.rs:436]: There are so many places this could go wrong in future code changes. Maybe it's fine to panic here, but I think an expect is clearer, and means we can skip the allow and comment. Something like:

```suggestion
            let id_to_remove = self.farthest_content_id.clone()
                .expect("farthest_content_id is always set immediately before this block");
```

File: trin-core/src/portalnet/storage.rs [hunk 3]
```diff
@@ -689,7 +696,9 @@ impl PortalStorage {
 
     /// Helper function for opening a RocksDB connection for the radius-constrained db.
     pub fn setup_rocksdb(node_id: NodeId) -> Result<rocksdb::DB, ContentStoreError> {
-        let mut data_path: PathBuf = get_data_dir(node_id);
+        let mut data_path: PathBuf = get_data_dir(node_id).map_err(|err| {
+            ContentStoreError::Database(format!("unable to get data dir: {:?}", err))
```
Comment (carver) [inline trin-core/src/portalnet/storage.rs:700]:
```suggestion
            ContentStoreError::Database(format!("unable to get data dir for rocksdb: {:?}", err))
```

for error uniqueness

File: trin-core/src/portalnet/storage.rs [hunk 4]
```diff
@@ -700,7 +709,9 @@ impl PortalStorage {
 
     /// Helper function for opening a SQLite connection.
     pub fn setup_sql(node_id: NodeId) -> Result<Pool<SqliteConnectionManager>, ContentStoreError> {
-        let mut data_path: PathBuf = get_data_dir(node_id);
+        let mut data_path: PathBuf = get_data_dir(node_id).map_err(|err| {
+            ContentStoreError::Database(format!("unable to get data dir: {:?}", err))
```
Comment (carver) [inline trin-core/src/portalnet/storage.rs:713]:
```suggestion
            ContentStoreError::Database(format!("unable to get data dir for sql: {:?}", err))
```

for error uniqueness

File: trin-core/src/utp/packets.rs [hunk 5]
```diff
@@ -142,7 +142,7 @@ impl PacketHeader {
 
     /// Returns the packet's type.
     pub fn get_type(&self) -> PacketType {
-        PacketType::try_from(self.type_ver >> 4).unwrap()
+        PacketType::try_from(self.type_ver >> 4).expect("Invalid packet type")
```
Comment (carver) [inline trin-core/src/utp/packets.rs:145]: This string is used already. For uniqueness:

```suggestion
        PacketType::try_from(self.type_ver >> 4).expect("Can't parse packet type")
```

File: trin-state/src/utils.rs [hunk 6]
```diff
@@ -50,6 +50,8 @@ impl ToU256 for BigInt {
 // distance function as defined at...
 // https://notes.ethereum.org/h58LZcqqRRuarxx4etOnGQ#Storage-Layout
 // todo: measure performance & optimize if necessary
+// naked wraps ok here because we're operating on static values
+#[allow(clippy::unwrap_used)]
```
Comment (carver) [inline trin-state/src/utils.rs:54]: Slight preference to move this down to the `unwrap()` usages, rather than allowing the whole method.

Also, between the overhead of adding allow, and the comment, I am thinking it might be better to generally prefer `expect`s, even when doing static parsing. Like:

```rs
   let modulo = BigInt::parse_bytes(&MODULO, 10).expect("Parse static MODULO always works");
   let mid = BigInt::parse_bytes(&MID, 10).expect("Parse static MID always works");
```

This kind of thing could even be added to the style guide as a recommendation.

**Review by carver — APPROVED** [review]
Love it! 🚢 

It's great how often unwraps could be replaced by `?`, or removed altogether.
